### PR TITLE
[v2] Blob Metadata Store

### PR DESCRIPTION
**Pull request by ian-shim**
## Why are these changes needed?
Set up v2 blob metadata store. Only implements few basic methods for now. 

## Checks

- [ ] I've made sure the lint is passing in this PR.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- [ ] I've checked the new test coverage and the coverage percentage didn't drop.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: core/data.go [hunk 1]
```diff
@@ -517,3 +518,27 @@ type ActiveReservation struct {
...
+	return BlobKey(b), nil
+}
+
+type BlobHeaderV2 struct {
+	BlobVersion    BlobVersion              `json:"version"`
+	QuorumIDs      []QuorumID               `json:"quorum_ids"`
+	BlobCommitment encoding.BlobCommitments `json:"commitments"`
+
+	PaymentMetadata `json:"payment_metadata"`
+}
```
Comment (ian-shim) [inline core/data.go:544]: These will be replaced by v2 core types in https://github.com/Layr-Labs/eigenda/pull/806

File: disperser/common/v2/blobstore/dynamo_store.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	}
+
+	return count, nil
+}
+
+func GenerateTableSchema(tableName string, readCapacityUnits int64, writeCapacityUnits int64) *dynamodb.CreateTableInput {
+	return &dynamodb.CreateTableInput{
+		AttributeDefinitions: []types.AttributeDefinition{
+			{
+				AttributeName: aws.String("PK"),
```
Comment (pschork) [inline disperser/common/v2/blobstore/dynamo_store.go:134]: Nice!

File: disperser/common/v2/blobstore/dynamo_store.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+)
+
+const (
+	StatusIndexName            = "StatusIndex"
+	OperatorDispersalIndexName = "OperatorDispersalIndex"
+	OperatorResponseIndexName  = "OperatorResponseIndex"
+
+	blobKeyPrefix     = "BlobKey#"
+	dispersalIDPrefix = "DispersalID#"
+	blobMetadataSK    = "BlobMetadata"
```
Comment (pschork) [inline disperser/common/v2/blobstore/dynamo_store.go:28]: So we're not using immutable design where blob metadata updates get a new unique sortable `SK` ie `BlobMetadata#UUID7` or `BlobMetadata#EpochSeconds`?

File: disperser/common/v2/blob.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+	core.BlobHeaderV2 `json:"blob_header"`
+
+	BlobKey    core.BlobKey `json:"blob_key"`
+	BlobStatus BlobStatus   `json:"blob_status"`
+	// Expiry is Unix timestamp of the blob expiry in seconds from epoch
+	Expiry      uint64 `json:"expiry"`
+	NumRetries  uint   `json:"num_retries"`
+	BlobSize    uint64 `json:"blob_size"`
+	RequestedAt uint64 `json:"requested_at"`
+}
```
Comment (cody-littley) [inline disperser/common/v2/blob.go:24]: Will this eventually contain information such as blob header and/or blob cert?

**Review by cody-littley — APPROVED** [review]

File: disperser/common/v2/blob.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ian-shim) [reply disperser/common/v2/blob.go:24]: good call

Comment (ian-shim) [reply disperser/common/v2/blob.go:24]: It already includes blob header. It represents an disperser internal record tracked in DB. Blob cert will be a separate type and we can probably reuse what's defined in `core/v2`

File: disperser/common/v2/blobstore/dynamo_store.go (hunk 3 above)
Comment (ian-shim) [reply disperser/common/v2/blobstore/dynamo_store.go:28]: yeah as discussed offline, some fields (BlobStatus) will be updated

File: core/data.go [hunk 5]
```diff
@@ -483,34 +484,53 @@ func (cb Bundles) FromEncodedBundles(eb EncodedBundles) (Bundles, error) {
...
-	binIndexBytes := make([]byte, 4)
-	binary.BigEndian.PutUint32(binIndexBytes, pm.BinIndex)
-	data = append(data, binIndexBytes...)
-
-	// Append CumulativePayment
-	paymentBytes := pm.CumulativePayment.Bytes()
-	data = append(data, paymentBytes...)
-
-	return crypto.Keccak256(data)
+func (pm *PaymentMetadata) Hash() ([32]byte, error) {
```
Comment (ian-shim) [inline core/data.go:498]: updated this hash function to make it compatible with solidity 
cc @hopeyen

File: disperser/common/v2/blobstore/dynamo_store.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return metadata, nil
+}
+
+// GetBlobMetadataByStatus returns all the metadata with the given status
+// Because this function scans the entire index, it should only be used for status with a limited number of items.
+func (s *blobMetadataStore) GetBlobMetadataByStatus(ctx context.Context, status v2.BlobStatus) ([]*v2.BlobMetadata, error) {
```
Comment (mooselumph) [inline disperser/common/v2/blobstore/dynamo_store.go:89]: Would there ever be a need for a batched version of this?

File: core/data.go [hunk 7]
```diff
@@ -483,34 +486,80 @@ func (cb Bundles) FromEncodedBundles(eb EncodedBundles) (Bundles, error) {
 
 // PaymentMetadata represents the header information for a blob
 type PaymentMetadata struct {
-	// Existing fields
-	AccountID string
+	// AccountID is the ETH account address for the payer
+	AccountID string `json:"account_id"`
```
Comment (mooselumph) [inline core/data.go:490]: Where is the json serialization used?

File: core/data.go [hunk 8]
```diff
@@ -483,34 +486,80 @@ func (cb Bundles) FromEncodedBundles(eb EncodedBundles) (Bundles, error) {
...
+
+	var hash [32]byte
+	hasher := sha3.NewLegacyKeccak256()
+	hasher.Write(bytes)
+	copy(hash[:], hasher.Sum(nil)[:32])
+
+	return hash, nil
+}
+
+func (pm *PaymentMetadata) MarshalDynamoDBAttributeValue() (types.AttributeValue, error) {
```
Comment (mooselumph) [inline core/data.go:538]: Does this belong in core? Is there any way to have this code in the blobstore?

File: core/v2/types.go [hunk 9]
```diff
@@ -59,7 +81,153 @@ func (b *BlobHeader) GetEncodingParams() (encoding.EncodingParams, error) {
 		NumChunks:   uint64(params.NumChunks),
 		ChunkLength: uint64(length),
 	}, nil
+}
+
+func (b *BlobHeader) BlobKey() (BlobKey, error) {
```
Comment (mooselumph) [inline core/v2/types.go:86]: Nice!

**Review by mooselumph — COMMENTED** [review]
looks good!

File: core/data.go (hunk 7 above)
Comment (ian-shim) [reply core/data.go:490]: It's used when we marshal/unmarshal the struct to/from json, i.e. in dataapi.
I've removed these for now

File: core/data.go (hunk 8 above)
Comment (ian-shim) [reply core/data.go:538]: They need to belong in the same package as where the type is defined. We can wrap and clone these types in blobstore but not sure if that's worth it

File: disperser/common/v2/blobstore/dynamo_store.go [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+)
+
+const (
+	StatusIndexName            = "StatusIndex"
+	OperatorDispersalIndexName = "OperatorDispersalIndex"
+	OperatorResponseIndexName  = "OperatorResponseIndex"
+
+	blobKeyPrefix     = "BlobKey#"
+	dispersalIDPrefix = "DispersalID#"
```
[the remark this answers is not in the document]
Comment (ian-shim) [reply disperser/common/v2/blobstore/dynamo_store.go]: It's used in a follow up PR, removed for now

File: disperser/common/v2/blobstore/dynamo_store.go (hunk 6 above)
Comment (ian-shim) [reply disperser/common/v2/blobstore/dynamo_store.go]: What do you mean by a batched version?

**Review by cody-littley — APPROVED** [review]

**Pull request merged**